addon/-private/system/class-meta.js

```diff
@@ -0,0 +1,57 @@
+import EmptyObject from 'ember-data/-private/system/empty-object';
+
+export default class ClassMeta {
+  constructor(store, modelName) {
+    this.modelName = modelName;
+    this._store = store;
+    this._modelClass = null;
+    this._idToRecord = null;
+    this._records = null;
+    this._metadata = null;
+  }
+
+  get idToRecord() {
+    return this._idToRecord || (this._idToRecord = new EmptyObject());
+  }
+
+  get records() {
+    return this._records || (this._records = []);
+  }
+
+  get metadata() {
+    return this._metadata || (this._metadata = new EmptyObject());
+  }
+
+  get modelClass() {
+    if (!this._modelClass) {
+      this._modelClass = this._store.modelFor(this.modelName);
+    }
+
+    return this._modelClass;
+  }
+
+  /* deprecated way of accessing modelClass */
+  get type() {
```

Comment:
@bmac I think the issue raised is that if there is (literally) no modelClass to resolve, that it will always be null, and ergo we will perpetually keep trying to resolve it with every "get" of type or modelClass.